doc/man/satd.1
.TH SATD 1 SAT
.SH NAME
satd \- Start the daemon that runs jobs queued for later execution.
.SH SYNOPSIS
.B satd
.RB [ \-f ]
.SH DESCRIPTION
.BR satd (1)
shall start the
.B sat
daemon, which is used for automatically executing jobs
queued for later execution.
.PP
Before a job is executed,
.BR satd (1)
will run the hook script if available, the hook action
will either be
.TP
.B expired
if the job run when scheduled to run, or
.TP
.B forced
if
.BR satr (1)
was used to run the job early.
.PP
After the job has run, the hook script is run again,
and is made sure that the hook script is not run any
time between these two runs. This time, the action is
either
.TP
.B failure
if the job was not executed successfully, including if
the job exited with any status other than zero, or
.TP
.B success
if the job was executed successfully and exited with
status zero.
.PP
When a job is removed using
.BR satrm (1),
the hook script is run with the action
.BR removed .
When a job is queued using
.BR sat (1),
the hook script is run with the action
.BR queued .
.PP
When the hook script runs, the first argument (not
counting the zeroth argument: the pathname of hook
script) is the action, the follow arguments is the
command line of the job. The environment will be set
to be identical to that of the job.
.SH OPTIONS
.TP
.B \-f
Run the daemon in the foreground.
.SH ENVIRONMENT
.TP
.B XDG_RUNTIME_DIR
This environment variable names the directory in which
interprocess communication related files are stored. If
unset or empty, /run is used.
.TP
.B SAT_HOOK_PATH
The pathname of the hook script to use. Does not have to
already exist. If not defined, $XDG_CONFIG_HOME/sat/hook
(if XDG_CONFIG_HOME is defined), $HOME/.config/sat/hook
(if HOME is defined), ~/.config/sat/hook (if the user has
a home and is not root), or /etc/sat/hook (otherwise) is
used.
.SH "SEE ALSO"
.BR sat (1),
.BR satq (1),
.BR satrm (1),
.BR satd (1)
.SH AUTHORS
Principal author, Mattias Andrée.  See the COPYING file for the full
list of authors.
.SH LICENSE
MIT/Expat License.
.SH BUGS
Please report bugs to https://github.com/maandree/sat/issues or to
maandree@member.fsf.org


NEWS
sat NEWS                                              -*- outline -*-

* Noteworthy changes in release 1.1 (2016-(01)Jan-01 UTC) [stable]

  sat does not use a daemon socket anymore,
  the actions of the commands are handled directly
  in the command rather than delegated. The daemon
  is let known via a SIGCHLD signal (not code added
  to satd, well barely any, only removed.)


* Noteworthy changes in release 1.0 (2016-(01)Jan-01 UTC) [stable]

  Initial release.



src/satd.c
/**
 * Copyright © 2015, 2016  Mattias Andrée <maandree@member.fsf.org>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */
#include "common.h"
#include "daemonise.h"



COMMAND("satd")
USAGE("[-f]")



/**
 * Create and lock the lock file, and its directory.
 * 
 * @return  A file descriptor to the lock file, -1 on error.
 */
static int
create_lock(void)
{
	const char *dir;
	char *path;
	char *p;
	pid_t pid = getpid();
	int fd = -1, saved_errno = 0;

	/* Create directory. */
	dir = getenv("XDG_RUNTIME_DIR"), dir = (dir ? dir : "/run");
	t (!(path = malloc(strlen(dir) * sizeof(char) + sizeof("/" PACKAGE "/lock"))));
	p = stpcpy(stpcpy(path, dir), "/" PACKAGE);
	t (mkdir(path, S_IRWXU) && (errno != EEXIST));

	/* Open file. */
	stpcpy(p, "/lock");
	t (fd = open(path, O_RDWR | O_CREAT /* but not O_EXCL or O_TRUNC */, S_IRUSR | S_IWUSR), fd == -1);

	/* Check that the daemon is not running, and mark it as running. */
	if (flock(fd, LOCK_EX | LOCK_NB)) {
		t (errno != EWOULDBLOCK);
		fprintf(stderr, "%s: the daemon is already in reading.\n", argv0);
		errno = 0;
		goto fail;
	}

	/* Store PID in the file. */
	/* Yes it is coming similar to a PID file, but this works if the started with -f. */
	t (pwrite(fd, &pid, sizeof(pid), (off_t)0) < (ssize_t)sizeof(pid));

	goto done;
fail:
	saved_errno = errno, close(fd), fd = -1;
done:
	free(path), errno = saved_errno;
	return fd;
}


/**
 * The sat daemon initialisation.
 * 
 * @param   argc  Any value in [0, 2] is accepted.
 * @param   argv  The name of the process, and -f if the process
 *                shall not be daemonised.
 * @return  0     The process was successful.
 * @return  1     The process failed queuing the job.
 * @return  2     User error, you do not know what you are doing.
 */
int
main(int argc, char *argv[])
{
	int state = -1, boot = -1, real = -1, lock = -1, foreground = 0;
	char *path = NULL;
	struct itimerspec spec;

	/* Parse command line. */
	if (argc > 0)  argv0 = argv[0];
	if (argc > 2)  usage();
	if ((argc == 2) && (!(foreground = !strcmp(argv[1], "-f"))))
		usage();

	/* Get hook-script pathname. */
	t (set_hookpath());

	/* Open/create lock file and state file. */
	GET_FD(lock,  LOCK_FILENO,  create_lock());
	GET_FD(state, STATE_FILENO, open_state(O_RDWR | O_CREAT, &path));

	/* Create timers. */
	GET_FD(boot, BOOT_FILENO, timerfd_create(CLOCK_BOOTTIME, 0));
	GET_FD(real, REAL_FILENO, timerfd_create(CLOCK_REALTIME, 0));

	/* Configure timers. */
	memset(&spec, 0, sizeof(spec));
	t (timerfd_settime(boot, TFD_TIMER_ABSTIME, &spec, NULL));
	t (timerfd_settime(real, TFD_TIMER_ABSTIME, &spec, NULL));

	/* Daemonise. */
	t (foreground ? 0 : daemonise("satd", /*DAEMONISE_KEEP_FDS | DAEMONISE_NEW_PID,*/ 3, 4, 5, 6, -1));

	/* Change to a process image without all this initialisation text. */
	execl(LIBEXECDIR "/" PACKAGE "/satd-diminished", argv0, path, NULL);

fail:
	if (errno)
		perror(argv0);
	free(path);
	close(state), close(boot), close(real), close(lock);
	undaemonise();
	return 1;
}

